File: 1/2020.09.12/25.Reverse Nodes in k-Group/python/solution.py
class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # recursively
        cur, count = head, 0
        while cur and count != k:
            cur = cur.next;
            count += 1
        if count == k:
            cur = self.reverseKGroup(cur, k);
            while count > 0:
                count -= 1
                tmp = head.next
                head.next = cur
                cur = head
                head = tmp
            head = cur
        return head

    def reverseKGroup2(self, head, k):
        # iteratively
        hair = ListNode(0)
        hair.next = head
        pre = hair
        while head:
            tail = pre
            for _ in range(k):
                tail = tail.next
                if not tail:
                    return hair.next
            nex = tail.next
            head, tail = self.reverse(head, tail)
            pre.next = head
            tail.next = nex
            pre = tail
            head = pre.next
        return hair.next

    def reverse(self, head, tail):
        prev = tail.next
        p = head
        while prev != tail:
            nex = p.next
            p.next = prev
            prev = p
            p = nex
        return tail, head

File: 1/2020.09.12/25.Reverse Nodes in k-Group/java/solution.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
		// recursively
		ListNode cur = head;
		int count = 0;
		while (cur != null && count != k ) {
			cur = cur.next;
			count++;
		}
		if (count == k) {
			cur = reverseKGroup(cur, k);
			while (count > 0) {
				count--;
				ListNode tmp = head.next;
				head.next = cur;
				cur = head;
				head = tmp;
			}
			head = cur;
		}
		return head;
	}
	
	public ListNode reverseKGroup2(ListNode head, int k) {
		// iteratively
		if (head == null || head.next == null || k == 1) {
			return head;
		}
		ListNode dummy = new ListNode(0);
		dummy.next = head;
		ListNode begin = dummy;
		int i = 0;
		while (head != null) {
			i++;
			if (i % k == 0) {
				begin = reverse(begin, head.next);
				head = begin.next;
			} else {
				head = head.next;
			}
		}
		return dummy.next;
	}

	private ListNode reverse(ListNode begin, ListNode end) {
		ListNode result = begin.next;
		ListNode cur = result;
		ListNode prev = begin;
		while (cur != end) {
			ListNode next = cur.next;
			cur.next = prev;
			prev = cur;
			cur = next;
		}
		begin.next = prev;
		result.next = end;
		return result;
	}
}
